engines/freescape/gfx.cpp

```diff
@@ -0,0 +1,153 @@
+	return surf;
+}
+
+bool Renderer::getRGBAt(uint8 index, uint8 &r, uint8 &g, uint8 &b) {
+
+	if (index == _keyColor)
+		return false;
+
+	if (index == 0) {
+		_palette->getRGBAt(0, r, g, b);
+		return true;
+	}
+
+	if (_renderMode == Common::kRenderAmiga || _renderMode == Common::kRenderAtariST) {
+		_palette->getRGBAt(index, r, g, b);
+		return true;
+	}
+
+	// assert(index-1 < _colorMap->size());
+	byte *entry = (*_colorMap)[index - 1];
+	uint8 color = 0;
+	uint8 acc = 1;
+	for (int i = 0; i < 4; i++) {
+		byte be = *entry;
+		if (be != 0 && be != 0xff) {
+			// TODO: fix colors for non-DOS releases
+			_palette->getRGBAt(index, r, g, b);
+			return true;
+		}
+
+		if (be == 0xff)
+			color = color + acc;
+
+		acc = acc << 1;
+		entry++;
+	}
+	assert(color < 16);
+	_palette->getRGBAt(color, r, g, b);
+	return true;
+}
+
+Common::Rect Renderer::viewport() const {
+	return _screenViewport;
+}
+
+void Renderer::computeScreenViewport() {
+	_screenViewport = Common::Rect(_screenW, _screenH);
+}
+
+Renderer *createRenderer(OSystem *system, int screenW, int screenH, Common::RenderMode renderMode) {
+	Common::String rendererConfig = ConfMan.get("renderer");
+	Graphics::PixelFormat pixelFormat = Graphics::PixelFormat(4, 8, 8, 8, 8, 24, 16, 8, 0);
+	Graphics::RendererType desiredRendererType = Graphics::kRendererTypeTinyGL;  // Graphics::parseRendererTypeCode(rendererConfig);
+	Graphics::RendererType matchingRendererType = Graphics::kRendererTypeTinyGL; // Graphics::getBestMatchingAvailableRendererType(desiredRendererType);
+
+	bool isAccelerated = 0; // matchingRendererType != Graphics::kRendererTypeTinyGL;
+
+	if (isAccelerated) {
+		initGraphics3d(screenW, screenH);
+	} else {
+		initGraphics(screenW, screenH, &pixelFormat);
+	}
+
+	/*#if defined(USE_OPENGL_GAME) || defined(USE_OPENGL_SHADERS) || defined(USE_GLES2)
+		bool backendCapableOpenGL = g_system->hasFeature(OSystem::kFeatureOpenGLForGame);
+	#endif
+
+	#if defined(USE_OPENGL_GAME)
+		// Check the OpenGL context actually supports shaders
+		if (backendCapableOpenGL && matchingRendererType == Graphics::kRendererTypeOpenGLShaders && !OpenGLContext.shadersSupported) {
+			matchingRendererType = Graphics::kRendererTypeOpenGL;
+		}
+	#endif*/
+
+	if (matchingRendererType != desiredRendererType && desiredRendererType != Graphics::kRendererTypeDefault) {
+		// Display a warning if unable to use the desired renderer
+		warning("Unable to create a '%s' renderer", rendererConfig.c_str());
+	}
+	/*
+	#if defined(USE_GLES2) || defined(USE_OPENGL_SHADERS)
+		if (backendCapableOpenGL && matchingRendererType == Graphics::kRendererTypeOpenGLShaders) {
+			return CreateGfxOpenGLShader(system);
+		}
+	#endif
+	#if defined(USE_OPENGL_GAME) && !defined(USE_GLES2)
+		if (backendCapableOpenGL && matchingRendererType == Graphics::kRendererTypeOpenGL) {
+			return CreateGfxOpenGL(system);
+		}
+	#endif*/
+	if (matchingRendererType == Graphics::kRendererTypeTinyGL) {
```

Comment:
you should allow compiler without tinygl

Reply:
Since this is the only rendered so far, it will be fine to trigger an error saying that this engine cannot run without TinyGL support?

Reply:
I would wrote more generic error, like no renderers enabled or something

Reply:
Done!